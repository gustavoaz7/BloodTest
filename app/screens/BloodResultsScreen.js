import React, { Component } from 'react';
import { View, FlatList, Alert, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';
import BloodResultItem from '../components/BloodResultItem';
import PendingWrapper from '../components/PendingWrapper';
import { isRejected } from '../redux/utils/status';
import { stateWithStatusType, stackNavigationType } from '../types';
import { SCREENS } from '../constants';

export default class BloodResultScreen extends Component {
  static propTypes = {
    loadBloodResults: PropTypes.func.isRequired,
    results: stateWithStatusType.isRequired,
    navigation: stackNavigationType.isRequired,
  };

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.loadResults();
  }

  loadResults = async () => {
    const { loadBloodResults } = this.props;
    loadBloodResults();
  };

  navigateTemplate = () => {
    this.props.navigation.navigate(SCREENS.TEMPLATE, { from: 'BloodResultsScreen' });
  };

  handleError() {
    Alert.alert(
      'Something went wrong :(',
      'There was a problem connecting to our servers.\n\nPlease try again in a few seconds.',
      [
        {
          text: 'Try again',
          onPress: this.loadResults,
        },
      ],
    );
  }

  keyExtractor(item) {
    return item.id;
  }

  renderItem = ({ item }) => {
    return <BloodResultItem item={item} onPress={this.navigateTemplate} />;
  };

  renderSeparator() {
    return <Separator />;
  }

  render() {
    const { results } = this.props;

    if (isRejected(results)) {
      this.handleError();
      return null;
    }
    return (
      <PendingWrapper value={results}>
        <FlatListUI
          data={results.value}
          renderItem={this.renderItem}
          keyExtractor={this.keyExtractor}
          ItemSeparatorComponent={this.renderSeparator}
          refreshing={false}
          onRefresh={this.loadResults}
          contentContainerStyle={styles.contentStyle}
        />
      </PendingWrapper>
    );
  }
}

export const FlatListUI = styled(FlatList)`
  background-color: 'rgba(0, 0, 0, 0.05)';
`;

const Separator = styled(View)`
  height: 12px;
`;

const styles = StyleSheet.create({
  contentStyle: {
    padding: 12,
  },
});
