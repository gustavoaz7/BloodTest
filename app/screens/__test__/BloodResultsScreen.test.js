import React from 'react';
import { shallow } from 'enzyme';
import BloodResultsScreen, { FlatListUI } from '../BloodResultsScreen';
import {
  LOAD_BLOOD_RESULTS_FULFILLED_STATE as fulfilledState,
  LOAD_BLOOD_RESULTS_PENDING_STATE as pendingState,
  LOAD_BLOOD_RESULTS_REJECTED_STATE as rejectedState,
} from '../../redux/BloodResults/test.data';

describe('BloodResultScreen', () => {
  const noop = () => {};

  describe('renders without crashing', () => {
    it('pending state', () => {
      shallow(<BloodResultsScreen {...pendingState} loadBloodResults={noop} />);
    });
    it('fulfilled state', () => {
      shallow(<BloodResultsScreen {...fulfilledState} loadBloodResults={noop} />);
    });
    it('rejected state', () => {
      shallow(<BloodResultsScreen {...rejectedState} loadBloodResults={noop} />);
    });
  });
  it('loads tests when mouting', async () => {
    const spy = jest.fn();
    shallow(<BloodResultsScreen {...fulfilledState} loadBloodResults={spy} />);

    expect(spy).toHaveBeenCalled();
  });

  it('reloads tests on `pull to refresh` ', async () => {
    const spy = jest.fn();
    const wrapper = shallow(<BloodResultsScreen {...fulfilledState} loadBloodResults={spy} />);

    spy.mockClear();

    wrapper
      .find(FlatListUI)
      .props()
      .onRefresh();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe('matches snapshot', () => {
    it('pending state', () => {
      const wrapper = shallow(<BloodResultsScreen {...pendingState} loadBloodResults={noop} />);

      expect(wrapper).toMatchSnapshot('pending');
    });

    it('rejected state', async () => {
      const wrapper = shallow(<BloodResultsScreen {...rejectedState} loadBloodResults={noop} />);

      expect(wrapper).toMatchSnapshot('rejected');
    });

    it('success state', async () => {
      const wrapper = shallow(<BloodResultsScreen {...fulfilledState} loadBloodResults={noop} />);

      expect(wrapper).toMatchSnapshot('success');
    });
  });
});
