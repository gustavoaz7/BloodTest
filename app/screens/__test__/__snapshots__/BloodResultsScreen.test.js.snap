// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BloodResultScreen matches snapshot pending state: pending 1`] = `
<PendingWrapper
  loadingProps={null}
  value={
    Object {
      "status": "PENDING",
      "value": null,
    }
  }
>
  <Styled(FlatList)
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={
      Object {
        "padding": 12,
      }
    }
    data={null}
    keyExtractor={[Function]}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
  />
</PendingWrapper>
`;

exports[`BloodResultScreen matches snapshot rejected state: rejected 1`] = `""`;

exports[`BloodResultScreen matches snapshot success state: success 1`] = `
<PendingWrapper
  loadingProps={null}
  value={
    Object {
      "status": "FULFILLED",
      "value": Array [
        Object {
          "email": "Berniece.Haley42@hotmail.com",
          "id": "0b82fed2-a091-4318-ba02-4d5e6192d6f7",
          "name": "Citlalli Kuhic",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.48,
            },
            Object {
              "time": "afternoon",
              "value": 0.39,
            },
            Object {
              "time": "evening",
              "value": 0.19,
            },
          ],
        },
        Object {
          "email": "Teagan_Ratke83@gmail.com",
          "id": "acf65e6a-1878-4cd9-bf6f-440dafd4fa18",
          "name": "Rachael Lehner",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.67,
            },
            Object {
              "time": "afternoon",
              "value": 0.31,
            },
            Object {
              "time": "evening",
              "value": 0.24,
            },
          ],
        },
        Object {
          "email": "Palma17@yahoo.com",
          "id": "ca285be1-0d6b-42f6-bfec-b3aeaa2aaf8f",
          "name": "Willa Quitzon",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.73,
            },
            Object {
              "time": "afternoon",
              "value": 0.19,
            },
            Object {
              "time": "evening",
              "value": 0.14,
            },
          ],
        },
        Object {
          "email": "Giuseppe19@hotmail.com",
          "id": "95eba822-a7b3-4938-8c94-005733701c1f",
          "name": "Miss Audie Moore",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.26,
            },
            Object {
              "time": "afternoon",
              "value": 0.09,
            },
            Object {
              "time": "evening",
              "value": 0.29,
            },
          ],
        },
        Object {
          "email": "Maeve_Jacobs14@yahoo.com",
          "id": "4fea89aa-bcf7-42b0-a5e8-202fc03a02fe",
          "name": "Yesenia Nicolas",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.78,
            },
            Object {
              "time": "afternoon",
              "value": 0.26,
            },
            Object {
              "time": "evening",
              "value": 0.24,
            },
          ],
        },
        Object {
          "email": "Johnson90@yahoo.com",
          "id": "17d11f05-2580-49b5-bb17-b786209c4792",
          "name": "Ms. Harmon Kunze",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.47,
            },
            Object {
              "time": "afternoon",
              "value": 0.09,
            },
            Object {
              "time": "evening",
              "value": 0.09,
            },
          ],
        },
      ],
    }
  }
>
  <Styled(FlatList)
    ItemSeparatorComponent={[Function]}
    contentContainerStyle={
      Object {
        "padding": 12,
      }
    }
    data={
      Array [
        Object {
          "email": "Berniece.Haley42@hotmail.com",
          "id": "0b82fed2-a091-4318-ba02-4d5e6192d6f7",
          "name": "Citlalli Kuhic",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.48,
            },
            Object {
              "time": "afternoon",
              "value": 0.39,
            },
            Object {
              "time": "evening",
              "value": 0.19,
            },
          ],
        },
        Object {
          "email": "Teagan_Ratke83@gmail.com",
          "id": "acf65e6a-1878-4cd9-bf6f-440dafd4fa18",
          "name": "Rachael Lehner",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.67,
            },
            Object {
              "time": "afternoon",
              "value": 0.31,
            },
            Object {
              "time": "evening",
              "value": 0.24,
            },
          ],
        },
        Object {
          "email": "Palma17@yahoo.com",
          "id": "ca285be1-0d6b-42f6-bfec-b3aeaa2aaf8f",
          "name": "Willa Quitzon",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.73,
            },
            Object {
              "time": "afternoon",
              "value": 0.19,
            },
            Object {
              "time": "evening",
              "value": 0.14,
            },
          ],
        },
        Object {
          "email": "Giuseppe19@hotmail.com",
          "id": "95eba822-a7b3-4938-8c94-005733701c1f",
          "name": "Miss Audie Moore",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.26,
            },
            Object {
              "time": "afternoon",
              "value": 0.09,
            },
            Object {
              "time": "evening",
              "value": 0.29,
            },
          ],
        },
        Object {
          "email": "Maeve_Jacobs14@yahoo.com",
          "id": "4fea89aa-bcf7-42b0-a5e8-202fc03a02fe",
          "name": "Yesenia Nicolas",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.78,
            },
            Object {
              "time": "afternoon",
              "value": 0.26,
            },
            Object {
              "time": "evening",
              "value": 0.24,
            },
          ],
        },
        Object {
          "email": "Johnson90@yahoo.com",
          "id": "17d11f05-2580-49b5-bb17-b786209c4792",
          "name": "Ms. Harmon Kunze",
          "samples": Array [
            Object {
              "time": "morning",
              "value": 0.47,
            },
            Object {
              "time": "afternoon",
              "value": 0.09,
            },
            Object {
              "time": "evening",
              "value": 0.09,
            },
          ],
        },
      ]
    }
    keyExtractor={[Function]}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
  />
</PendingWrapper>
`;
